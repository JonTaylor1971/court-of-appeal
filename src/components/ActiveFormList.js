import React, { Component } from 'react';
import '../forms/active.forms.css';
import DefaultService from '../service/default.service.js';
import CaseList from '../components/CaseList.js';
import _ from 'lodash';
import update from 'immutability-helper';

class ActiveFormList extends Component {

    constructor(props) {
        super(props);
        this.service = props.service; 
        this.state = {
            fetch: props.fetch !== 'false',
            cases: [],
            displayMyCasesEmptyLabel:true
        };
        this.fetchCases = this.fetchCases.bind(this);
        this.updateCases = this.updateCases.bind(this);
    }

    componentDidMount() {
        let window = this.element.ownerDocument.defaultView;
        if (this.service == null) { this.service = new DefaultService(window); }
        if (this.state.fetch) { this.fetchCases(); }
    }

    fetchCases() {
        this.service.getMyCases({}, (data) => {             
            if (data && data.cases && data.cases.length > 0) {
                this.setState({ 
                    cases:data.cases.slice(0, 5).map(function(item) { 
                        return {
                            id:item.id,
                            parties: item.data.appellants[0].name + ' / ' + item.data.respondents[0].name,
                            status: item.status,
                            modified: item.modified,
                            data: item.data
                        };
                    }),
                    displayMyCasesEmptyLabel: false 
                });
            } else {
                this.setState({
                    cases:[],
                    displayMyCasesEmptyLabel: true
                })
            }
        });
    }

    /**
     * Update the cases in the state so we don't need to call the api to get the updated cases after a successful update.
     */
    updateCases(data, id) {
        let cases = this.state.cases;
        _.forEach(cases, (doc) => {
            if (doc.id === id) {
                doc.data = data;
            }
        });
        this.setState(update(this.state, {cases: {$set: cases}}));
    }

    render() {
        return (
            <div id="active-forms" className="form-section" ref={ (element)=> {this.element = element }}>
                <div>
                    <h3 style={{ display:'inline-block' }}>Document Reminders</h3>
                    <div style={{ display:'inline-block', float:'right', marginTop:'15px' }}>
                        <a href="/my-applications.html" className="btn btn-primary round-borders">View All Documents</a>
                    </div>
                </div>
                <CaseList
                    cases={this.state.cases}                   
                    service={this.service}
                    updateCases={this.updateCases}
                />
                <div id="my-cases-empty-label" style={{ display:this.state.displayMyCasesEmptyLabel?'block':'none' }}>
                        No open cases found
                </div>
            </div>
        )
    }
}
export default ActiveFormList;