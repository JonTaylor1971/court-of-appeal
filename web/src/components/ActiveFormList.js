import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import '../MyDocuments.css';
import CaseList from '../components/CaseList.js';

class ActiveFormList extends Component {

    render() {
        return (
            <div id="active-forms" className="form-section" ref={ (element)=> {this.element = element }}>
                <div className="not-printable">
                    <h3 style={{ display:'inline-block' }}>Document Reminders</h3>

                </div>
                <CaseList
                    cases={this.props.cases}
                    service={this.props.service}
                    updateCases={this.props.updateCases}
                    history={this.props.history}
                />
                <div id="my-cases-empty-label" style={{ display:this.props.displayMyCasesEmptyLabel ? 'block' : 'none' }}>
                        No open cases found
                </div>
                <div>
                    <Link to={process.env.PUBLIC_URL + '/my-documents'} className="btn btn-primary round-borders">View All Documents</Link>
                </div>
            </div>
        )
    }
}
export default ActiveFormList;
