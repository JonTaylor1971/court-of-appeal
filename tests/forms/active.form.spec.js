require('../support/fake.dom');
import React from 'react';
import renderer from "react-test-renderer";
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ActiveFormList from '../../src/components/ActiveFormList';
let LocalServer = require('../support/local.server');

configure({ adapter: new Adapter() });

describe('Active forms section', function() {

    let apiServer;

    let activeFormList;
    let cases = [
        { id:1501, status:'draft', modified:'2018-03-27T16:15:54Z', data:{appellants: [{name:'Bruce'}], respondents:[{name:'Clark'}]} }
    ];

    beforeEach(function(done) {
        apiServer = new LocalServer((request, response)=> {  
            if (request.url == '/api/cases' && request.method == 'GET') {                
                response.write( JSON.stringify({ cases:cases })); 
                response.end();
            }
            else {
                response.statusCode = 200;
                response.write('ko');
                response.end(); 
            }
        });
        apiServer.start(done);
         activeFormList = mount(
            <ActiveFormList fetch="false" service={{ getMyCases: (form, callback) => { callback( {cases: [
                        { id:1501, status:'draft', modified:'2018-03-27T16:15:54Z', data:{appellants: [{name:'Bruce'}], respondents:[{name:'Clark'}]} }
                    ]});} }}/>,
        ).instance();
    });

    afterEach(function(done) {
        apiServer.stop(done);
    });
    
    test('transforms the data for the list', function(done) {
        activeFormList.fetchCases();
        setTimeout(()=> {
            expect(activeFormList.state.cases).toEqual([
                { 
                    id:1501, status:'draft', modified:'2018-03-27T16:15:54Z', parties:'Bruce / Clark' ,
                    data:{appellants:[{name:'Bruce'}], respondents:[{name:'Clark'}]}
                }
            ]);
            done();
        }, 100);        
    });

});

test ('renders full list of documents ', ()=> {
    let list = renderer.create(
        <ActiveFormList fetch="false" service={{ getMyCases: (form, callback) => { callback( {cases: [
                    { id:1501, status:'draft', modified:'2018-03-27T16:15:54Z', data:{appellants: [{name:'Bruce'}], respondents:[{name:'Clark'}]} }
                ]});} }}/>,
    );
    let tree = list.toJSON();
    expect(tree).toMatchSnapshot();
    let instance = list.getInstance();
    instance.fetchCases();
    expect(instance.state.displayMyCasesEmptyLabel).toEqual(false);


});

test( 'renders empty list of documents ', ()=> {

    let emptylist = renderer.create(
        <ActiveFormList fetch="false" service={{ getMyCases: (form, callback) => { callback( { cases: [] }); } }}/>
    );
    let tree = emptylist.toJSON();
    expect(tree).toMatchSnapshot();

    let instance = emptylist.getInstance();
    instance.fetchCases();
    expect(instance.state.displayMyCasesEmptyLabel).toEqual(true);
});