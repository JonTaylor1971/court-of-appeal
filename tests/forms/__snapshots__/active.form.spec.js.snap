// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active forms section display empty mention when no cases are found 1`] = `
<div
  className="form-section"
  id="active-forms"
>
  <div
    className="not-printable"
  >
    <h3
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      Document Reminders
    </h3>
    <div
      style={
        Object {
          "display": "inline-block",
          "float": "right",
          "marginTop": "15px",
        }
      }
    >
      <a
        className="btn btn-primary round-borders"
        href="/my-applications.html"
      >
        View All Documents
      </a>
    </div>
  </div>
  <div
    id="case-list"
    onClick={[Function]}
  >
    <table
      className="not-printable"
      id="my-cases"
    >
      <thead>
        <tr
          className="header"
        >
          <td>
            Edit
          </td>
          <td>
            File #
          </td>
          <td>
            Parties
          </td>
          <td>
            Status
          </td>
          <td
            className="text-right"
          >
            Recently Modified
          </td>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      className="modal"
      id="editFormModal not-printable"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className=" modal-title "
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Edit Form 2
      </div>
      <div
        className="modal-content not-printable"
      >
        <div
          className="form-section"
          style={
            Object {
              "display": undefined,
            }
          }
        >
          <div
            className="button-bar not-printable"
          >
            <div>
              <button
                className="btn btn-warning"
                id="back"
                onClick={[Function]}
              >
                &lt; 
                Back
              </button>
            </div>
            <div>
              <button
                className="btn btn-success"
                id="draft"
                onClick={[Function]}
              >
                Save as Draft
              </button>
            </div>
            <div
              data-tip=""
            >
              <button
                className="btn btn-success"
                disabled={true}
                id="preview"
                onClick={[Function]}
              >
                Preview &gt;
              </button>
            </div>
            <div
              className="__react_component_tooltip place-top type-dark "
              data-id="tooltip"
              id={undefined}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal"
      id="viewFormModal"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="modal-title  not-printable"
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Preview Form 2
      </div>
      <div
        className="modal-content"
      >
        <div
          className="form-section"
          style={
            Object {
              "display": undefined,
            }
          }
        >
          <div
            className="button-bar not-printable"
          >
            <div>
              <button
                className="btn btn-warning"
                id="back"
                onClick={[Function]}
              >
                &lt; 
                Back to editing
              </button>
            </div>
            <div>
              <button
                className="btn btn-success"
                id="print"
                onClick={[Function]}
              >
                Print
              </button>
            </div>
            <div
              className="__react_component_tooltip place-top type-dark "
              data-id="tooltip"
              id={undefined}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal not-printable"
      id="dataLossWarning"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="modal-title warning"
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Warning!
      </div>
      <div
        className="modal-content"
      >
        <div>
          All changes will be lost!
        </div>
        <div
          className="button-bar not-printable"
        >
          <div>
            <button
              className="btn btn-warning"
              id="back"
              onClick={[Function]}
            >
              &lt; 
              Go back so I can save the form as draft.
            </button>
          </div>
          <div>
            <button
              className="btn btn-success"
              id="submit"
              onClick={[Function]}
            >
              Continue anyway.  I don't care about the data.
            </button>
          </div>
          <div
            className="__react_component_tooltip place-top type-dark "
            data-id="tooltip"
            id={undefined}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="my-cases-empty-label"
    style={
      Object {
        "display": "block",
      }
    }
  >
    No open cases found
  </div>
</div>
`;

exports[`Active forms section hides the empty mention when cases are found 1`] = `
<div
  className="form-section"
  id="active-forms"
>
  <div
    className="not-printable"
  >
    <h3
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      Document Reminders
    </h3>
    <div
      style={
        Object {
          "display": "inline-block",
          "float": "right",
          "marginTop": "15px",
        }
      }
    >
      <a
        className="btn btn-primary round-borders"
        href="/my-applications.html"
      >
        View All Documents
      </a>
    </div>
  </div>
  <div
    id="case-list"
    onClick={[Function]}
  >
    <table
      className="not-printable"
      id="my-cases"
    >
      <thead>
        <tr
          className="header"
        >
          <td>
            Edit
          </td>
          <td>
            File #
          </td>
          <td>
            Parties
          </td>
          <td>
            Status
          </td>
          <td
            className="text-right"
          >
            Recently Modified
          </td>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      className="modal"
      id="editFormModal not-printable"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className=" modal-title "
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Edit Form 2
      </div>
      <div
        className="modal-content not-printable"
      >
        <div
          className="form-section"
          style={
            Object {
              "display": undefined,
            }
          }
        >
          <div
            className="button-bar not-printable"
          >
            <div>
              <button
                className="btn btn-warning"
                id="back"
                onClick={[Function]}
              >
                &lt; 
                Back
              </button>
            </div>
            <div>
              <button
                className="btn btn-success"
                id="draft"
                onClick={[Function]}
              >
                Save as Draft
              </button>
            </div>
            <div
              data-tip=""
            >
              <button
                className="btn btn-success"
                disabled={true}
                id="preview"
                onClick={[Function]}
              >
                Preview &gt;
              </button>
            </div>
            <div
              className="__react_component_tooltip place-top type-dark "
              data-id="tooltip"
              id={undefined}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal"
      id="viewFormModal"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="modal-title  not-printable"
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Preview Form 2
      </div>
      <div
        className="modal-content"
      >
        <div
          className="form-section"
          style={
            Object {
              "display": undefined,
            }
          }
        >
          <div
            className="button-bar not-printable"
          >
            <div>
              <button
                className="btn btn-warning"
                id="back"
                onClick={[Function]}
              >
                &lt; 
                Back to editing
              </button>
            </div>
            <div>
              <button
                className="btn btn-success"
                id="print"
                onClick={[Function]}
              >
                Print
              </button>
            </div>
            <div
              className="__react_component_tooltip place-top type-dark "
              data-id="tooltip"
              id={undefined}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal not-printable"
      id="dataLossWarning"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="modal-title warning"
      >
        <span
          id="close-modal"
          onClick={[Function]}
        >
          ×
        </span>
        Warning!
      </div>
      <div
        className="modal-content"
      >
        <div>
          All changes will be lost!
        </div>
        <div
          className="button-bar not-printable"
        >
          <div>
            <button
              className="btn btn-warning"
              id="back"
              onClick={[Function]}
            >
              &lt; 
              Go back so I can save the form as draft.
            </button>
          </div>
          <div>
            <button
              className="btn btn-success"
              id="submit"
              onClick={[Function]}
            >
              Continue anyway.  I don't care about the data.
            </button>
          </div>
          <div
            className="__react_component_tooltip place-top type-dark "
            data-id="tooltip"
            id={undefined}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="my-cases-empty-label"
    style={
      Object {
        "display": "block",
      }
    }
  >
    No open cases found
  </div>
</div>
`;
